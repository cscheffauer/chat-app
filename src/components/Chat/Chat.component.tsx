import React, { FormEvent, KeyboardEvent, useState, useEffect, useRef } from 'react'

import Message from './Message.component'
import './Chat.scss';

import { MessageType } from '../../models/chat.model'

interface Props {
    sendMessage: (message: String) => void,
    editMessage: (messageid: String, text: String) => void,
    deleteMessage: (messageid: String) => void,
    messages: Array<MessageType>,
    userid: String,
}

const Chat = ({ sendMessage, editMessage, messages, ...PropsFromParent }: Props) => {
    const [message, setmessage] = useState("");
    const [editmode, seteditmode] = useState(false);
    const [messageidtoedit, setmessageidtoedit] = useState("" as String);
    const [originalmessage, setoriginalmessage] = useState('' as String);

    const messageInput = useRef<HTMLInputElement>(null);        //ref for messageInput
    const scrollArea = useRef<HTMLDivElement>(null);        //ref for scrollArea

    useEffect(() => {
        if (messageInput.current !== null) {
            messageInput.current.focus();           //focus on the messageInput after messages changed
        }
        if (scrollArea.current !== null) {
            scrollArea.current.scrollTop = scrollArea.current.scrollHeight;     //to scroll to the bottom after messages changed
        }
    }, [messages]);


    const handleChange = (event: FormEvent<HTMLInputElement>) => {
        event.preventDefault();
        setmessage(event.currentTarget.value)
    }

    const handleSubmit = (event: FormEvent | KeyboardEvent) => {
        event.preventDefault();
        if (message.length > 0) {
            if (editmode) {
                if (originalmessage !== message) {
                    editMessage(messageidtoedit, message);
                    cancelEditMode();
                }
            } else {
                sendMessage(message);
                setmessage("");
            }
        }
    }

    const switchToEditMode = (messageid: String, messagetext: String) => {
        setmessageidtoedit(messageid);
        setoriginalmessage(messagetext);
        setmessage(messagetext as string);
        seteditmode(true);
        if (messageInput.current !== null) {
            messageInput.current.focus();           //focus on the messageInput to edit the message
        }
    }

    const cancelEditMode = () => {
        seteditmode(false);
        setmessage("");
    }

    return (
        <div className="chatWindow">
            <div ref={scrollArea} className="chatMessages">
                {
                    messages.map((message, i) => <Message key={i} message={message} switchToEditMode={switchToEditMode} {...PropsFromParent} />)
                }
            </div>
            <div className="chatInput">
                <form onSubmit={handleSubmit}>
                    <input ref={messageInput} placeholder={"Message"} onChange={handleChange} onBlur={cancelEditMode} value={message} type="text"></input>
                    {editmode && <span>âœŽ</span>}
                </form>
            </div>
        </div>

    )
}

export default Chat
