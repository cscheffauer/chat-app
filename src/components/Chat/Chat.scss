.chatWindow {
    height: 84vh;
    background-color: white;
    display: flex;
    flex-direction: column;
    align-content: space-between;
    .chatMessages {
        flex:1;
        overflow-y: auto;
        .message {
            margin-top: 2vh;
            margin-left: 3vh;
        }
    }
    .chatBox {
        height: 8vh;
        display: table;
        form {
            display: table-cell;
            vertical-align: middle;
            text-align: center;
            input {
                border-radius: 5px;
                border: 1px solid #a3a3a3;
                font-size: 1.2em;
                height: 5vh;
                width: 94vw;
                box-sizing: border-box;
                padding: 12px;
            }
        }
    }
}