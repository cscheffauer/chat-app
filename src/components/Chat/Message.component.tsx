import React from 'react';
import Linkify from 'react-linkify';
import './Chat.scss';

interface Props {
    message: {
        sent: Date,
        username: String,
        text: String,
    }
}

const Message = ({ message }: Props) => {
    const sentDate = new Date(message.sent);
    const sent = sentDate.getHours() + ":" + sentDate.getMinutes();
    return (
        <div className="message">
            <p><span style={{ color: "#00002c", fontWeight: 500 }}>{message.username}</span><span style={{ color: "#9da2aa", marginLeft: 10 }}>{sent}</span> </p>
            <Linkify>
                <p style={{ lineHeight: '150%', color: message.username === "Meetingbot" ? 'grey' : 'black' }}>{message.text}</p>
            </Linkify>
        </div >
    )
}

export default Message
