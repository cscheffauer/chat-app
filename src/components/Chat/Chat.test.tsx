import React from 'react';
import { shallow } from 'enzyme';
import Chat from './Chat.component';

describe('Component / Chat', () => {
    const sendMessageMock = () => { };
    const messagesMock = [{ username: 'Max', text: 'Hallo', sent: new Date() }, { username: 'John', text: 'Hi', sent: new Date() }, { username: 'Jeffrey', text: "What's up?", sent: new Date() }];
    const ChatWrapper = shallow(<Chat sendMessage={sendMessageMock} messages={messagesMock} />);

    it('expect to render Chat component', () => {
        expect(ChatWrapper).toMatchSnapshot();
    })
})


