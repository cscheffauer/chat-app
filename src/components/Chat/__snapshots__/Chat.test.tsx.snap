// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component / Chat expect to render Chat component 1`] = `ShallowWrapper {}`;
